### PR TITLE
docs(core/config): add warning for deprecated options

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Add two new fields `deprecated` and `depreationMsg` to `RenovateOptionBase` interface, and use them to display warning message for deprecated config options.
- Add new function `genDeprecationMsg` similar to `genExperimentalMsg` 

## Context
https://github.com/renovatebot/renovate/pull/28104#discussion_r1542827012

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Cannot be done only with deprecationMsg?

**Review by secustor — DISMISSED** [review]

File: lib/config/options/index.ts [hunk 1]
```diff
@@ -2065,6 +2065,8 @@ const options: RenovateOptions[] = [
     description: 'Branch name template.',
     type: 'string',
     default: '{{{branchPrefix}}}{{{additionalBranchPrefix}}}{{{branchTopic}}}',
+    deprecationMsg:
+      'Direct editing of branchName is now deprecated. Please edit branchPrefix, additionalBranchPrefix, or branchTopic instead`',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:2069]:
```suggestion
      'Directly editing the branchName is deprecated. Please use branchPrefix, additionalBranchPrefix, or branchTopic.`',
```

Can we use `monospace` styling in the deprecation message? If yes, we should add the `monospace` styling to the config option names.

File: lib/config/options/index.ts [hunk 2]
```diff
@@ -2088,6 +2090,8 @@ const options: RenovateOptions[] = [
     type: 'string',
     default:
       '{{{commitMessagePrefix}}} {{{commitMessageAction}}} {{{commitMessageTopic}}} {{{commitMessageExtra}}} {{{commitMessageSuffix}}}',
+    deprecationMsg:
+      'Direct editing of commitMessage is now deprecated. Please edit commitMessage subcomponents instead.',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:2094]:
```suggestion
      'Direct editing of the commitMessage is deprecated. Please edit commitMessage subcomponents instead.',
```

The message should probably list the subcomponents, or link to a docs entry that explains the correct procedure.

File: lib/config/options/index.ts [hunk 3]
```diff
@@ -2158,9 +2162,11 @@ const options: RenovateOptions[] = [
   {
     name: 'prTitle',
     description:
-      'Pull Request title template (deprecated). Inherits from `commitMessage` if null.',
+      'Pull Request title template. Inherits from `commitMessage` if null.',
     type: 'string',
     default: null,
+    deprecationMsg:
+      'Direct editing of prTitle is now deprecated. Please edit commitMessage subcomponents instead as they will be passed through to prTitle.',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:2169]:
```suggestion
      'Direct editing of prTitle is deprecated. Please edit commitMessage subcomponents instead as they will be passed through to prTitle.',
```

Same comment about listing or linking to the docs for the replacement subcomponents here.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts (hunk 2 above)
Comment (RahulGautamSingh) [reply lib/config/options/index.ts:2094]: https://docs.renovatebot.com/configuration-options/#commitmessage

There is already a depreciation warning for `commitMessage`, which lists the subcomponents. Maybe we can shorten the `depreciationMsg` instead or there will be a double warning for the same reason. What do you suggest will be a better way to handle this?

File: lib/config/options/index.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/config/options/index.ts:2069]: I am not sure about this. @viceice wdyt??

Comment (viceice) [reply lib/config/options/index.ts:2069]: we can probably use it for the config key words

File: lib/config/options/index.ts (hunk 2 above)
Comment (viceice) [reply lib/config/options/index.ts:2094]: remove the existing warning in favor of this.

maybe we should add an optional deprecated related options field where we list alternatives.

File: lib/config/types.ts [hunk 4]
```diff
@@ -434,6 +434,8 @@ export interface RenovateOptionBase {
   experimentalIssues?: number[];
 
   advancedUse?: boolean;
+
+  deprecationMsg?: string;
```
Comment (viceice) [inline lib/config/types.ts:438]: add tsdoc that this is for docs

File: lib/config/options/index.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/config/options/index.ts:2069]: @HonkingGoose I am not sure how to add monospace styling here, can you help

**Comment by RahulGautamSingh** [comment]
<details>
<summary>prTitle Before</summary>
image

</details>

**Comment by RahulGautamSingh** [comment]
Before and After images of the docs after the changes:

**`branchName`**

image
---

**`commitMessage`**

image

---

**`prTitle`**

image

**Review by viceice — APPROVED** [review]

**Review by HonkingGoose — APPROVED** [review]

**Review by secustor — APPROVED** [review]

**Pull request merged**